dask/tests/test_task_spec.py

```diff
@@ -491,7 +494,10 @@ def test_inline():
 
     t = Task("key-1", identity, [TaskRef("a"), TaskRef("b")])
     inlined = t.inline({"a": DataNode(None, "a"), "b": DataNode(None, "b")})
-    assert inlined() == ["a", "b"]
+    assert inlined() == [DataNode(None, "a"), DataNode(None, "b")]
+    t = Task("key-1", func, TaskRef("a"), TaskRef("b"))
+    inlined = t.inline({"a": DataNode(None, "a"), "b": DataNode(None, "b")})
+    assert inlined() == "a-b"
```

Comment:
This test shows probably best what the impact of this change will be. Passing nested objects, whether they are references or actual tasks is no longer supported and will silently produce odd results.

We may want to implement a test-only validation step that still recurses into data structures to ensure we're not messing this up accidentally but ultimately this will safe a lot of computation time during runtime.

i.e. if I genuinely want to pass a list, this would have to look like...

```python
In [1]: from dask._task_spec import TaskRef, Task

In [2]: def func(a):
   ...:     return "-".join(a)
   ...:
   ...: def aslist(*args):
   ...:     return list(args)
   ...:
   ...: task_with_nested_refs = Task("key", func, Task(None, aslist, TaskRef("foo"), TaskRef("bar")))

In [3]: task_with_nested_refs.dependencies
Out[3]: frozenset({'bar', 'foo'})

In [4]: task_with_nested_refs({"foo": "foo", "bar": "bar"})
Out[4]: 'foo-bar'
```

This is why I had the SequenceOfTasks / DictOfTasks classes initially (though they were also poorly implemented. I think it makes sense to add dedicated functions/classes for these operations to avoid having to construct the above stuff manually all the time

cc @hendrikmakait